File: hard/0004-median-of-two-sorted-arrays/python/solution.py
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        median = 0

        if len(nums1) == 0 and len(nums2) == 0:
            return median

        total_size = len(nums1) + len(nums2)
        result = []
        while len(nums1) != 0 or len(nums2) != 0:
            if len(nums1) == 0:
                result.extend(nums2)
                break

            if len(nums2) == 0:
                result.extend(nums1)
                break

            if nums1[0] < nums2[0]:
                el = nums1.pop(0)
                result.append(el)
            else:
                el = nums2.pop(0)
                result.append(el)

            # TODO(dio): Optimisation - once we are past
            # the midpoint of the result_array, we don't
            # need to carry on - we can just stop there

        if total_size % 2 == 0:
            median = result[total_size // 2] + result[(total_size // 2) - 1]
            median /= 2
        else:
            median = result[total_size // 2]
        return median

File: hard/0004-median-of-two-sorted-arrays/go/solution.go
package solution

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	finalArr := make([]int, len(nums1)+len(nums2))
	finalArrIndex := 0
	medianIndexes := make([]int, 2)

	if len(finalArr)%2 == 0 {
		// two indexes
		medianIndexes[0] = (len(finalArr) / 2) - 1
		medianIndexes[1] = len(finalArr) / 2
	} else {
		medianIndexes[0] = len(finalArr) / 2
	}

	longestArr, shortestArr := longestArr(nums1, nums2)

	i := 0
	j := 0
	for i < len(longestArr) {
		for j < len(shortestArr) {
			if i == len(longestArr) {
				for ; j < len(shortestArr); j++ {
					finalArr[finalArrIndex] = shortestArr[j]
					finalArrIndex++
				}
				break
			}

			if longestArr[i] < shortestArr[j] {
				finalArr[finalArrIndex] = longestArr[i]
				i++
			} else {
				finalArr[finalArrIndex] = shortestArr[j]
				j++
			}
			finalArrIndex++
		}

		if j == len(shortestArr) {
			for ; i < len(longestArr); i++ {
				finalArr[finalArrIndex] = longestArr[i]
				finalArrIndex++
			}
			break
		}
	}

	if len(finalArr)%2 == 0 {
		return float64((finalArr[medianIndexes[0]] + finalArr[medianIndexes[1]])) / 2.0
	}
	return float64(finalArr[medianIndexes[0]])
}

func longestArr(nums1 []int, nums2 []int) ([]int, []int) {
	if len(nums1) > len(nums2) {
		return nums1, nums2
	}

	return nums2, nums1
}
